### PR TITLE
Feature enumerable

### PR DESCRIPTION
**Pull request by mariancristu**
Enumerable Project

 [ x ] All methods asked done
 [ x ] Checked with Rubocop
 
 All methods are contained in **enumerable.rb** file.

 Thank you and have a nice day!

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                        (!datatype.empty? && (datatype[0].class == Regexp) && x.to_s.match(datatype[0]).nil?) ||
+                        (datatype.empty? && (!yield(x) || x.nil?))
+      end
+    else
+      my_each do |x, y|
+        return false if !datatype.empty? || (datatype.empty? && !yield(x, y))
+      end
+    end
+    true
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [ ] `my_all?` is identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_all?(5) # returns this error

    # TypeError:
       class or module required
     # ./enumerable.rb:60:in `is_a?'
     # ./enumerable.rb:60:in `block in my_all?'
     # ./enumerable.rb:16:in `my_each'
     # ./enumerable.rb:59:in `my_all?'
  ```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                       (!datatype.empty? && (datatype[0].class == Regexp) && !x.to_s.match(datatype[0]).nil?) ||
+                       (datatype.empty? && yield(x))
+      end
+    else
+      my_each do |x, y|
+        return true if datatype.empty? || yield(x, y)
+      end
+    end
+    false
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [ ] `my_any?` is identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_any?('cat') # returns this error

   #  TypeError:
       class or module required
     # ./enumerable.rb:78:in `is_a?'
     # ./enumerable.rb:78:in `block in my_any?'
     # ./enumerable.rb:16:in `my_each'
     # ./enumerable.rb:77:in `my_any?'
  ```

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+  def my_count(*value)
+    return to_a.length if !block_given? && value.empty?
+
+    count = 0
+    to_a.my_each { |x| count += 1 if x == value[0] } unless value.empty?
+    to_a.my_each { |x, y| count += 1 if yield(x, y) } if block_given?
+    count
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [ ] should have working `my_none?` **(covered by automated tests)** identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_none?(5) # returns this error

    # TypeError:
       class or module required
     # ./enumerable.rb:78:in `is_a?'
     # ./enumerable.rb:78:in `block in my_any?'
     # ./enumerable.rb:16:in `my_each'
     # ./enumerable.rb:77:in `my_any?'
     # ./enumerable.rb:91:in `my_none?'
  ```

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      to_a.my_each do |x|
+        memo = memo.nil? ? x : memo.send(value[-1], x)
+      end
+      return memo
+    end
+    to_a.my_each do |x|
+      memo = memo.nil? ? x : yield(memo, x)
+    end
+    memo
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [ ] `my_inject` is identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } == range.inject(4) { |prod, n| prod * n } # false
  ```
* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) == range.my_inject(2, :*) # false
  ```

**Comment by mariancristu** [comment]
Thank you for the review. It was really explicit! I liked very much!

**Pull request merged**